src/abi/substrate.rs

```diff
@@ -1,284 +1,459 @@
+            } else {
+                None
+            }
+        })
+        .collect();
+    let contract_name = ns.contracts[contract_no].name.clone();
+    let storage = Layout::Struct(StructLayout::new(contract_name, fields));
 
-#[derive(Deserialize, Serialize)]
-struct Storage {
-    #[serde(rename = "struct")]
-    structs: StorageStruct,
-}
+    let f_to_constructor = |f: &Function| -> ConstructorSpec<PortableForm> {
+        let payable = matches!(f.mutability, ast::Mutability::Payable(_));
+        let args = f
+            .params
+            .iter()
+            .map(|p| {
+                let ty = resolve_ast(&p.ty, ns, &mut registry);
 
-#[derive(Deserialize, Serialize)]
-struct StorageStruct {
-    fields: Vec<StorageLayout>,
-}
+                let path = registry.get(ty).unwrap().path().clone();
+                let spec = TypeSpec::new(ty.into(), path);
 
-#[derive(Deserialize, Serialize)]
-struct StorageLayout {
-    name: String,
-    layout: LayoutField,
-}
+                MessageParamSpec::new(p.name_as_str().to_string())
+                    .of_type(spec)
+                    .done()
+            })
+            .collect::<Vec<MessageParamSpec<PortableForm>>>();
+
+        ConstructorSpec::from_label("new".to_string())
+            .selector(f.selector().try_into().unwrap())
+            .payable(payable)
+            .args(args)
+            .docs(vec![render(&f.tags).as_str()])
+            .returns(ReturnTypeSpec::new(None))
+            .done()
+    };
 
-#[derive(Deserialize, Serialize)]
-struct LayoutField {
-    cell: LayoutFieldCell,
-}
+    let constructors = ns.contracts[contract_no]
+        .functions
+        .iter()
+        .filter_map(|i| {
+            // include functions of type constructor
+            let f = &ns.functions[*i];
+            if f.is_constructor() {
+                Some(f)
+            } else {
+                None
+            }
+        })
+        .chain(
+            // include default constructor if exists
+            ns.contracts[contract_no]
+                .default_constructor
+                .as_ref()
+                .map(|(e, _)| e),
+        )
+        .map(f_to_constructor)
+        .collect::<Vec<ConstructorSpec<PortableForm>>>();
 
-#[derive(Deserialize, Serialize)]
-struct LayoutFieldCell {
-    key: String,
-    ty: usize,
-}
+    let conflicting_names = non_unique_function_names(contract_no, ns);
 
-/// Create a new registry and create new entries. Note that the registry is
-/// accessed by number, and the first entry is 1, not 0.
-impl Abi {
-    /// Add a type to the list unless already present
-    fn register_ty(&mut self, ty: Type) -> usize {
-        match self.types.iter().position(|t| *t == ty) {
-            Some(i) => i + 1,
-            None => {
-                self.types.push(ty);
-
-                self.types.len()
+    let f_to_message = |f: &Function| -> MessageSpec<PortableForm> {
```

Comment:
`f_to_message` is not great. How about `function_to_message` or simply `message_spec`